src/popups/checkout_option.rs

```
@@ -0,0 +1,245 @@
+	app::Environment,
+	keys::{key_match, SharedKeyConfig},
+	queue::Queue,
+	strings,
+	ui::{self, style::SharedTheme},
+};
+use anyhow::{Ok, Result};
+use asyncgit::sync::branch::checkout_remote_branch;
+use asyncgit::sync::status::discard_status;
+use asyncgit::sync::{checkout_branch, BranchInfo, RepoPath};
+use crossterm::event::Event;
+use ratatui::{
+	layout::{Alignment, Rect},
+	text::{Line, Span},
+	widgets::{Block, Borders, Clear, Paragraph},
+	Frame,
+};
+
+pub struct CheckoutOptionPopup {
+	queue: Queue,
+	repo: RepoPath,
+	branch: Option<BranchInfo>,
+	option: CheckoutOptions,
+	visible: bool,
+	key_config: SharedKeyConfig,
+	theme: SharedTheme,
+}
+
+impl CheckoutOptionPopup {
+	///
+	pub fn new(env: &Environment) -> Self {
+		Self {
+			queue: env.queue.clone(),
+			repo: env.repo.borrow().clone(),
+			branch: None,
+			option: CheckoutOptions::Unchange,
+			visible: false,
+			key_config: env.key_config.clone(),
+			theme: env.theme.clone(),
+		}
+	}
+
+	fn get_text(&self, _width: u16) -> Vec<Line<'_>> {
+		let mut txt: Vec<Line> = Vec::with_capacity(10);
+
+		txt.push(Line::from(vec![
+			Span::styled(
+				String::from("Switch to: "),
+				self.theme.text(true, false),
+			),
+			Span::styled(
+				self.branch.as_ref().expect("No branch").name.clone(),
+				self.theme.commit_hash(false),
+			),
+		]));
+
+		let (kind_name, kind_desc) = self.option.to_string_pair();
+
+		txt.push(Line::from(vec![
+			Span::styled(
+				String::from("How: "),
+				self.theme.text(true, false),
+			),
+			Span::styled(kind_name, self.theme.text(true, true)),
+			Span::styled(kind_desc, self.theme.text(true, false)),
+		]));
+
+		txt
+	}
+
+	///
+	pub fn open(&mut self, branch: BranchInfo) -> Result<()> {
+		self.show()?;
+
+		self.branch = Some(branch);
+
+		Ok(())
+	}
+
+	fn checkout(&self) -> Result<()> {
+		if let Some(branch) = &self.branch {
+			if branch.is_local() {
+				checkout_branch(
+					&self.repo,
+					&self.branch.as_ref().expect("No branch").name,
```

why do we want to panic here instead of returning the error upwards?

---

Something’s wrong here, I realized I should be using the branch returned by `let Some(branch)`.